### PR TITLE
Added the endpoint GetParty in LookupController.cs.

### PR DESCRIPTION
**Pull request by Sondre-B**
## Description
Added endpoint for looking up party by partyId.

## Related Issue(s)
- #224 

## Verification
- [ ] **Your** code builds clean without any errors or warnings
- [ ] Manual testing done (required)
- [ ] Relevant automated test added (if you find this hard, leave it and we'll help out)
- [ ] All tests run green

## Documentation
- [ ] User documentation is updated with a separate linked PR in [altinn-studio-docs.](https://github.com/Altinn/altinn-studio-docs) (if applicable)

### CODE REVIEW

File: src/Altinn.AccessManagement/Altinn.AccessManagement.csproj [hunk 1]
```diff
@@ -12,7 +12,7 @@
     <PackageReference Include="AutoMapper.Extensions.Microsoft.DependencyInjection" Version="12.0.0" />
     <PackageReference Include="Azure.Extensions.AspNetCore.Configuration.Secrets" Version="1.2.2" />
     <PackageReference Include="Azure.Identity" Version="1.8.0" />
-    <PackageReference Include="JWTCookieAuthentication" Version="2.4.2" />
+    <PackageReference Include="JWTCookieAuthentication" Version="3.0.0-alpha2" />
```
Comment (TheTechArch) [inline src/Altinn.AccessManagement/Altinn.AccessManagement.csproj:15]: Må få husket å publisere denne til 3.0.0

File: src/Altinn.AccessManagement/Controllers/LookupController.cs [hunk 2]
```diff
@@ -63,5 +64,45 @@ public async Task<ActionResult<PartyExternal>> GetOrganisation(string orgNummer)
...
+        public async Task<ActionResult<PartyExternal>> GetParty(int partyId)
+        {           
+            try
+            {
+                int userId = AuthenticationHelper.GetUserId(HttpContext);
+                List<Party> partyList = await _register.GetPartiesForUser(userId);
+               
+                if (partyList.Count > 0)
+                {
+                    foreach (Party party in partyList)
```
Comment (TheTechArch) [inline src/Altinn.AccessManagement/Controllers/LookupController.cs:85]: Linq is probably little faster here

File: src/Altinn.AccessManagement/appsettings.json [hunk 3]
```diff
@@ -64,5 +64,15 @@
     "SBLBaseAdress": "https://at22.altinn.cloud/",
     "AccessManagementApplicationHostName": "localhost",
     "DisableCsrfCheck": false
+  },
+  "OidcProviders": {
```
Comment (TheTechArch) [inline src/Altinn.AccessManagement/appsettings.json:68]: Forgotten config from earlier?

File: src/Altinn.AccessManagement/Controllers/LookupController.cs (hunk 2 above)
Comment (jonkjetiloye) [reply src/Altinn.AccessManagement/Controllers/LookupController.cs:85]: Also, the reporteelist is hierarchical so need to check for child parties. See PartiesWrapper.ValidateSelectedParty in authorization for example: [PartiesWrapper](https://github.com/Altinn/altinn-authorization/blob/main/src/Authorization/Services/Implementation/PartiesWrapper.cs) which does the same but only returns a bool and not the party

Comment (Sondre-B) [reply src/Altinn.AccessManagement/Controllers/LookupController.cs]: Moved logit to servicelayer and implemented support for subunits.

File: src/Altinn.AccessManagement.Core/Services/Interfaces/IRegister.cs [hunk 4]
```diff
@@ -8,10 +8,26 @@ namespace Altinn.AccessManagement.Core.Services.Interfaces
...
+        /// <summary>
+        /// Gets a party based on partyId
+        /// </summary>
+        /// <param name="partyId">The id of party to look for</param>
+        /// <param name="userId">The id of authenticated user</param>
+        /// <returns>Party based on partyId</returns>
+        public Task<Party> GetPartyForPartyId(int partyId, int userId);
+
+        /// <summary>
+        /// Gets a Party based on userId
```
Comment (jonkjetiloye) [inline src/Altinn.AccessManagement.Core/Services/Interfaces/IRegister.cs:26]: The returns documentation is way better that the summary

File: src/Altinn.AccessManagement.Core/Services/RegisterService.cs [hunk 5]
```diff
@@ -30,5 +29,50 @@ public Task<Party> GetOrganisation(string organisationNumber)
         {
             return _partyClient.LookupPartyBySSNOrOrgNo(organisationNumber);
         }
+
+        /// <inheritdoc/>
+        public Task<Party> GetPartyForPartyId(int partyId, int userId)
+        {
+            return _partyClient.GetPartyAsync(partyId);
```
[the remark this answers is not in the document]
Comment (jonkjetiloye) [reply src/Altinn.AccessManagement.Core/Services/RegisterService.cs:36]: So this is an unauthenticated party retrieval?

File: src/Altinn.AccessManagement.Core/Services/RegisterService.cs [hunk 6]
```diff
@@ -30,5 +29,50 @@ public Task<Party> GetOrganisation(string organisationNumber)
...
+        }
+
+        /// <inheritdoc/>
+        public async Task<Party> GetPartiesForUser(int userId, int partyId)
```
Comment (jonkjetiloye) [inline src/Altinn.AccessManagement.Core/Services/RegisterService.cs:40]: Plural name returns single entity. 

I also feel this is not a registry function. This is an authorization/accessmanagement function getting a reportee from the users reporteelist.

Comment (jonkjetiloye) [reply src/Altinn.AccessManagement.Core/Services/RegisterService.cs:40]: It should also probably be cached. Consider moving implementation to the ContextRetrievalService maybe?

File: src/Altinn.AccessManagement.Core/Services/ContextRetrievalService.cs [hunk 7]
```diff
@@ -216,5 +216,64 @@ public async Task<List<ServiceResource>> GetResources()
...
+            if (_memoryCache.TryGetValue(cacheKey, out List<Party> partyListFromCache))
+            {
+                return partyListFromCache;
+            }
+
+            List<Party> partyList = await _partiesClient.GetPartiesForUserAsync(userId);
+
+            var cacheEntryOptions = new MemoryCacheEntryOptions()
+              .SetPriority(CacheItemPriority.High)
+              .SetAbsoluteExpiration(new TimeSpan(0, _cacheConfig.ResourceRegistryResourceCacheTimeout, 0));
```
Comment (jonkjetiloye) [inline src/Altinn.AccessManagement.Core/Services/ContextRetrievalService.cs:272]: can either reuse _cacheConfig.PartyCacheTimeout or create a new for this reporteelist lookup. either is at least better than using the timeout for ResourceRegistryResourceCacheTimeout

File: test/Altinn.AccessManagement.Tests/Controllers/LookupControllerTest.cs [hunk 8]
```diff
@@ -106,10 +107,130 @@ private HttpClient GetTestClient()
...
+            // Assert
+            Assert.Equal(HttpStatusCode.OK, response.StatusCode);
+            AssertionUtil.AssertPartyEqual(expectedParty, actualParty);
+        }
+
+        /// <summary>
+        /// Test case: GetParty for partyId that does not correspond to a party in partylist for authenticated user.
+        /// Expected: GetParty returns 404 not found.
+        /// </summary>
+        /// <returns>404 not found</returns>
```
Comment (jonkjetiloye) [inline test/Altinn.AccessManagement.Tests/Controllers/LookupControllerTest.cs:204]: documented expected response and Assert does not match

**Review by jonkjetiloye — APPROVED** [review]

**Pull request merged**